### DISCUSSION TITLE
433Mhz module

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1194] [Q&A]
I tried to write the 433 MHz module, but unfortunately I encountered several problems and after more than 20 tests and outputs I could not solve it:

The problem is the monitor pin :
```javascript
new Monitor({
     pin: 22,
     mode: Digital.Input,
     edge: Monitor.Falling | Monitor.Rising
})
```

I think the monitor pin does not send accurate and correct information, because I wrote the exact same code in C, the code works for ESP32, but I use the monitor with javascript, the information is not correct.

GPIO Config C code :
```c

gpio_config_t data_pin_config = {
          .intr_type = GPIO_INTR_ANYEDGE,
          .mode = GPIO_MODE_INPUT,
          .pin_bit_mask = GPIO_SEL_22, // GPIO_NUM_22 (SEL) DATA PIN!
          .pull_up_en = GPIO_PULLUP_DISABLE,
          .pull_down_en = GPIO_PULLDOWN_DISABLE
 };

 gpio_config(&data_pin_config);

 // Attach the interrupt handler
gpio_install_isr_service(ESP_INTR_FLAG_EDGE);
gpio_isr_handler_add(GPIO_NUM_22, data_interrupt_handler, NULL);  // GPIO_NUM_22 DATA PIN!
```

What did I miss or what did I do wrong in the javascript settings that the 433Mhz receiver code works in C but not in javascript?

**Comment by phoddie** [comment]
I don't have experience working with a "433Mhz receiver" so I'm unsure what you are trying to do.

From the code, it looks like you want to have a callback invoked when pin 22 changes state (from low to high and from high to low). Is that correct?

You say it doesn't work, but what does that mean? For example, is the callback never invoked, is it invoked too often, is it invoked sometimes but missing some expected callbacks, etc? 

Also, how far apart in time are the expected callbacks?

**Reply by salarizadi** [reply]
Pay attention to the following data, after pressing one of the keys of the 433 MHz remote control, the 433 MHz module will send us the output 0 and 1, which we will convert into the time rate through the monitoring pin and according to the received...

Sample 0, 1 :
```
01010010100111110101100110001010011010011010110010101010010100111100010101100 // First click
101010010101111110100101100001010011011011010110010101010010100111100010101100 // Second click
10101000010111111010011110000101001101001100010001000101001010110001010101000001000 // Third click
0010110001011100101111010101010001010100101101011101011010100001000101101110110110101000110101010 // Fourth click
```

Javascript code :

> I'm sorry that the code is not clean, it is being tested intensively and accurately, all parts should be checked carefully and then the code will be cleaned :)

```javascript
const Protocols  = [
    [
        350,
        [1, 31],
        [1, 3],
        [3, 1],
        false
    ],    // protocol 1
    [650, [1, 10], [1, 2], [2, 1], false],    // protocol 2
    [100, [30, 71], [4, 11], [9, 6], false],  // protocol 3
    [380, [1, 6], [1, 3], [3, 1], false],     // protocol 4
    [500, [6, 14], [1, 2], [2, 1], false],    // protocol 5
    [450, [23, 1], [1, 2], [2, 1], true],     // protocol 6 (HT6P20B)
    [150, [2, 62], [1, 6], [6, 1], false],    // protocol 7 (HS2303-PT, i. e. used in AUKEY Remote)
    [200, [3, 130], [7, 16], [3, 16], false], // protocol 8 Conrad RS-200 RX
    [200, [130, 7], [16, 7], [16, 3], true],  // protocol 9 Conrad RS-200 TX
    [365, [18, 1], [3, 1], [1, 3], true],     // protocol 10 (1ByOne Doorbell)
    [270, [36, 1], [1, 2], [2, 1], true],     // protocol 11 (HT12E)
    [320, [36, 1], [1, 2], [2, 1], true]      // protocol 12 (SM5212)
];
const numProto = Protocols.length / Protocols[0].length;
let timings = [];
let RCSWITCH_MAX_CHANGES = 67
let nReceivedValue = 0;
let nReceivedBitlength = 0;
let nReceivedDelay = 0;
let nReceivedProtocol = 0;
let nReceiveTolerance = 60;
let nSeparationLimit = 4300;

class RCSwitch extends Monitor {

    #Structure = {
        init : false,
        timer: false,
        type : 0x2,
        scanning: false
    }
    #Raw = []

    lastTime    = 0
    changeCount = 0
    repeatCount = 0

    constructor ( options ) {
        if ( options.method === 0x0 ) {
            super({
                pin: options.pin,
                mode: Digital.Input,
                edge: Monitor.Falling | Monitor.Rising
            })
            this.onChanged = this.#receiver
        } else if ( options.method === 0x1 ) {

        }
    }

    transmit ( raw ) {}

    #receiver ( ) {
        let time = Time.microseconds;
        let duration = time - this.lastTime;

        if (duration > nSeparationLimit) {
            if (this.diff(duration, timings[0]) < 200) {
                this.repeatCount++;
                if (this.repeatCount == 2) {
                    for(let i = 1; i <= numProto; i++) {
                        trace(`\n Proto ${i}`)
                        if (this.#receiveProtocol(i, this.changeCount)) {
                            trace(`\nProtocol : ${i}, Code : ${nReceivedValue}, Bit : ${nReceivedBitlength}, Delay : ${nReceivedDelay}`)
                            break;
                        }
                    }
                    this.repeatCount = 0;
                }
            }
            this.changeCount = 0;
        }

        // detect overflow
        if (this.changeCount >= RCSWITCH_MAX_CHANGES) {
            this.changeCount = 0;
            this.repeatCount = 0;
        }

        timings[this.changeCount++] = duration;
        this.lastTime = time;
    }

    #receiveProtocol ( p ) {
        let pro = Protocols[p-1];

        const syncFactor = {
            high: pro[1][0],
            low : pro[1][1]
        };
        const zero = {
            high: pro[2][0],
            low : pro[2][1]
        };
        const one = {
            high: pro[3][0],
            low : pro[3][1]
        };
        const invertedSignal = pro[4]

        trace(`\n received times ${JSON.stringify(timings)}`)
        trace(`\n length times ${timings.length}`)
        trace(`\n changeCount ${this.changeCount}`)
        trace(`\n syncFactor : ${JSON.stringify(syncFactor)}`)
        trace(`\n zero : ${JSON.stringify(zero)}`)
        trace(`\n one : ${JSON.stringify(one)}`)
        trace(`\n invertedSignal : ${invertedSignal}`)

        let code = 0;
        //Assuming the longer pulse length is the pulse captured in timings[0]
        let syncLengthInPulses = ((syncFactor.low) > (syncFactor.high)) ? (syncFactor.low) : (syncFactor.high);
        let delay = timings[0] / syncLengthInPulses;
        let delayTolerance = delay * nReceiveTolerance / 100;

        let firstDataTiming = invertedSignal ? 2 : 1;

        for (let i = firstDataTiming; i < this.changeCount - 1; i += 2) {
            code <<= 1;
            if (
                this.diff(timings[i], delay * zero.high) < delayTolerance &&
                this.diff(timings[i + 1], delay * zero.low) < delayTolerance
            ) {
                // zero
            } else if (
                this.diff(timings[i], delay * one.high) < delayTolerance &&
                this.diff(timings[i + 1], delay * one.low) < delayTolerance
            ) {
                // one
                code |= 1;
            } else {
                // Failed
                trace(`\n Failed`)
                return false;
            }
        }

        if (this.changeCount > 7) {    // ignore very short transmissions: no device sends them, so this must be noise
            nReceivedValue = code;
            nReceivedBitlength = (this.changeCount - 1) / 2;
            nReceivedDelay = delay;
            nReceivedProtocol = p;
            return true;
        }

        return false;
    }

    die ( ) {}

    diff(A, B) {
        return Math.abs(A - B)
    }

}

RCSwitch.Receiver    = 0x0
RCSwitch.Transmitter = 0x1
RCSwitch.Decimal = 0x2
RCSwitch.Hex     = 0x3

Object.freeze(RCSwitch.prototype)

export default RCSwitch
```

As I said, using this formula in C language is not a problem, but in JavaScript, the information is not the same and accurate. Notice the output below :

The data you saw was sent from only one switch channel (switch A). Data is sent with 01, but converted to duration.

```
// First click :
 Proto 1
 received times [11000,1002,998,1000,1000,1000,141,1859,1000,1000,1000,1000,1000,1000,1000,68,932,1000,1000,1000,1000,1000,1000,1000,1000,1000,1005,64,931,1000,72,1928,141,859,1000,1000,1000,1005,78,1058,859,1000]
 length times 42
 changeCount 37
 syncFactor : {"high":1,"low":31}
 zero : {"high":1,"low":3}
 one : {"high":3,"low":1}
 invertedSignal : false
 Failed
 Proto 2
 received times [11000,1002,998,1000,1000,1000,141,1859,1000,1000,1000,1000,1000,1000,1000,68,932,1000,1000,1000,1000,1000,1000,1000,1000,1000,1005,64,931,1000,72,1928,141,859,1000,1000,1000,1005,78,1058,859,1000]
 length times 42
 changeCount 37
 syncFactor : {"high":1,"low":10}
 zero : {"high":1,"low":2}
 one : {"high":2,"low":1}
 invertedSignal : false
 Failed

Second click :
 Proto 1
 received times [11002,1000,998,141,859,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,141,1000,859,1000,1000,1000,1000,1000,1000,1000,1000,1000,141,1000,859,1000,141,1859,1000,1000,1000,1000,141,78,1058,859,1000]
 length times 42
 changeCount 37
 syncFactor : {"high":1,"low":31}
 zero : {"high":1,"low":3}
 one : {"high":3,"low":1}
 invertedSignal : false
 Failed
 Proto 2
 received times [11002,1000,998,141,859,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,141,1000,859,1000,1000,1000,1000,1000,1000,1000,1000,1000,141,1000,859,1000,141,1859,1000,1000,1000,1000,141,78,1058,859,1000]
 length times 42
 changeCount 37
 syncFactor : {"high":1,"low":10}
 zero : {"high":1,"low":2}
 one : {"high":2,"low":1}
 invertedSignal : false
 Failed

// Third click
 Proto 1
 received times [10859,1141,1000,861,998,141,1859,1000,1000,1000,1000,1000,1000,72,1069,859,1000,1000,1000,141,859,1000,1000,1000,1000,1141,864,995,1000,71,1929,1000,1000,1000,1000,1000,1000,141,78,1058,859,1000]
 length times 42
 changeCount 36
 syncFactor : {"high":1,"low":31}
 zero : {"high":1,"low":3}
 one : {"high":3,"low":1}
 invertedSignal : false
 Failed
 Proto 2
 received times [10859,1141,1000,861,998,141,1859,1000,1000,1000,1000,1000,1000,72,1069,859,1000,1000,1000,141,859,1000,1000,1000,1000,1141,864,995,1000,71,1929,1000,1000,1000,1000,1000,1000,141,78,1058,859,1000]
 length times 42
 changeCount 36
 syncFactor : {"high":1,"low":10}
 zero : {"high":1,"low":2}
 one : {"high":2,"low":1}
 invertedSignal : false
 Failed

// Fourth click
 Proto 1
 received times [11000,1003,997,1000,1000,1000,1000,1000,1000,1000,90,1051,859,1000,1000,1000,1000,1000,1000,1000,1000,1141,1000,859,1000,1000,1000,1000,1000,1000,1000,1000,1141,859,1000,1859,1000,141,78,1058,859,1000]
 length times 42
 changeCount 35
 syncFactor : {"high":1,"low":31}
 zero : {"high":1,"low":3}
 one : {"high":3,"low":1}
 invertedSignal : false
 Failed
 Proto 2
 received times [11000,1003,997,1000,1000,1000,1000,1000,1000,1000,90,1051,859,1000,1000,1000,1000,1000,1000,1000,1000,1141,1000,859,1000,1000,1000,1000,1000,1000,1000,1000,1141,859,1000,1859,1000,141,78,1058,859,1000]
 length times 42
 changeCount 35
 syncFactor : {"high":1,"low":10}
 zero : {"high":1,"low":2}
 one : {"high":2,"low":1}
 invertedSignal : false
 Failed
```

**Reply by salarizadi** [reply]
I wrote the code from here:

> It works perfectly correctly and without problems, it converts and shows the code sent from the remote correctly

https://github.com/DominikPalo/esp32-rf-receiver/blob/master/main/esp32_rf_receiver.h
https://github.com/DominikPalo/esp32-rf-receiver/blob/master/main/esp32_rf_receiver.c

Check it out, maybe you can find the problem.

**Reply by phoddie** [reply]
Thank you for the extensive background!

My guess is that the JavaScript code in #receiver is taking enough time that it causes some callbacks to be missed. A see a couple of possible solutions:

- Put the monitor into a separate Worker that just records the time values to a `SharedArrayBuffer` and leaves the processing to the main task. That has a good chance of catching all the callbacks.
- Use the RMT module. I think it is kind of designed for just this kind of situation. It is [documented here](https://www.moddable.com/documentation/pins/pins#rmt). There is are [IR Send and IR Receive examples](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/pins/rmt) by @stc1988 that use the RMT module that may be helpful.

**Reply by salarizadi** [reply]
I don't think that information will be lost because of these codes. If you pay attention to the example of the 01 codes, before I wrote the complete code, I tested the monitoring and only used trace(this.read()), then I pasted the 01 codes and saw that the output is not the same.

Before writing these codes, I used RMT first, but it was not received correctly, maybe I did not do the settings correctly, I need to collect more information.

**Reply by phoddie** [reply]
Are you asking for help determining the differences between the original C code and your JavaScript version? That's a lot of code to review.. Here's the first problem I noticed:

```js
const numProto = Protocols.length / Protocols[0].length;
```

That sets `numProto` to 2.4. I think you just need this:

```js
const numProto = Protocols.length;
```

**Reply by salarizadi** [reply]
Well done, 1 problem was found, but it still has a problem

**Reply by salarizadi** [reply]
I turned on the IR sensor module at the same time as the 433 and as you know the IR sensor works with the RMT and the 433 module works with the digital monitor but here comes a question why when I click on one of the IR remotes The monitor.onChange function is called.

At first I thought I was wrong, then I doubted and checked the pins, but they have nothing to do with each other and the IR sensor is set on RMT channel 0 and pin 33 and module 433 on pin 22.

not interesting? Do you have an answer to this problem?

**Reply by phoddie** [reply]
> Well done, 1 problem was found...

Great!

>...but it still has a problem

Sure. There's a lot of fresh code there. There could well be other issues. I don't have your hardware to reproduce the problem, so the best I can do is scan the code. But, please understand that I have other commitments on my time. I cannot provide on-demand debugging too.

> At first I thought I was wrong, then I doubted and checked the pins, but they have nothing to do with each other and the IR sensor is set on RMT channel 0 and pin 33 and module 433 on pin 22.

The digital monitor [installs a separate ISR with the ESP-IDF for each pin](https://github.com/Moddable-OpenSource/moddable/blob/787bf7ccbf69aa1f9dc42cb44cf50ae33f2b1ac1/modules/pins/digital/monitor/esp32/modMonitor.c#L142C28-L142C45). There should be no cross-talk at that level of the software.

**Reply by salarizadi** [reply]
Yes, it's true that we all have work time commitments, but sometimes we need to talk to identify and fix Moddable problems. I think the problem we have now is from the monitoring settings, I will definitely check the monitoring more :)

**Reply by phoddie** [reply]
> ...sometimes we need to talk to identify and fix Moddable problems

Absolutely.

>  I think the problem we have now is from the monitoring settings

If there's a problem in the monitor, I'd be happy to take a look. I just need a way to reproduce the problem, a pointer to a bug in the code, or a proposed patch.

**Reply by salarizadi** [reply]
I will definitely check it out. You know that I really like Moddable and it has helped me a lot in my projects. It's fast and convenient, especially since it can be programmed with JavaScript and I can solve the problem ;)

**Reply by linfan68** [reply]
@salarizadi in general reading zeros and ones with software (especially with js code) is a bad idea… IR data rate is very slow (1200 bps) and 433 can be as fast as 9600 or even higher.

I think RMT is the way to go, we have fairly good experiences using RMT to emulate and decode  UART and some custom protocols.

**Reply by salarizadi** [reply]
@linfan68 Thank you for your guidance, you pointed out a good point. I started with RMT, but unfortunately I could not write and find detailed settings for it. If you have an idea, be sure to suggest it with a code sample and clear information.

**Reply by phoddie** [reply]
@linfan68 - thank you for sharing your experience. I agree that the RMT is likely the best approach here.

@salarizadi – This is a forum for **discussion**. @linfan68 took the time to share his experience. There is no obligation in this discussion to provide sample code or detailed explanations.

Regarding the RMT, there is documentation and example code (links above). Those have been successfully used by other developers. I suggest starting with one of the examples. If you encounter questions along the way, please ask them.

**Reply by louisvangeldrop** [reply]
@linfan68 - RMT only works with ESP32 devices. With the support of nRF MCU's it is now possible to build Temperature-Humidity sensor based on DHT22 that reads the info once a minute and transmits the result e.g. with RF433 or BLE using a CR1632 battery. Reading a DHT22 requires reading falls/rises within the 25 -75 microseconds range.  It would help if the onChanged function also returns level and ticks in microseconds.
